pkg/api/kptfile/v1alpha2/types.go

```diff
@@ -0,0 +1,213 @@
+	// FastForward fails without updating if the local package was modified
+	// since it was fetched.
+	FastForward UpdateStrategyType = "fast-forward"
+	// ForceDeleteReplace wipes all local changes to the package.
+	ForceDeleteReplace UpdateStrategyType = "force-delete-replace"
+)
+
+// Upstream is a user-specified upstream locator for a package.
+type Upstream struct {
+	// Type is the type of origin.
+	Type OriginType `yaml:"type,omitempty"`
+
+	// Git is the locator for a package stored on Git.
+	Git *Git `yaml:"git,omitempty"`
+
+	// UpdateStrategy declares how a package will be updated from upstream.
+	UpdateStrategy UpdateStrategyType `yaml:"updateStrategy,omitempty"`
+}
+
+// UpstreamLock is a resolved locator for the last fetch of the package.
+type UpstreamLock struct {
+	// Type is the type of origin.
+	Type OriginType `yaml:"type,omitempty"`
+
+	// Git is the locator for a package stored on Git.
+	Git *Git `yaml:"git,omitempty"`
+}
+
+// Git contains information on the origin of packages fetched from a git repository.
+type Git struct {
+	// Repo is the git repository the package was cloned from.  e.g. https://
+	Repo string `yaml:"repo,omitempty"`
+
+	// Directory is the sub directory of the git repository that the package was cloned from
+	Directory string `yaml:"directory,omitempty"`
+
+	// Ref can be a Git branch, tag, or commit ID.
+	Ref string `yaml:"ref,omitempty"`
+
+	// Commit is the git commit that the package was fetched at
+	Commit string `yaml:"commit,omitempty"`
+}
+
+// PackageInfo contains information such as license, documentation, etc.
+// These fields are not consumed by any functionality in kpt and are simply passed through.
+type PackageInfo struct {
+	// URL is the location of the package.  e.g. https://github.com/example/com
+	URL string `yaml:"url,omitempty"`
+
+	// Email is the email of the package maintainer
+	Email string `yaml:"email,omitempty"`
+
+	// License is the package license
+	License string `yaml:"license,omitempty"`
+
+	// Version is a logical package version
+	Version string `yaml:"version,omitempty"`
+
+	// Tags enables humans and tools to attach arbitrary package metadata.
+	Tags []string `yaml:"tags,omitempty"`
+
+	// Man is the path to documentation about the package.
+	Man string `yaml:"man,omitempty"`
+
+	// Description contains a short description of the package.
+	Description string `yaml:"description,omitempty"`
+}
+
+// Subpackages declares a local or remote subpackage.
+type Subpackage struct {
+	// Name of the immediate subdirectory relative to this Kptfile where the suppackage
+	// either exists (local subpackages) or will be fetched to (remote subpckages).
+	// This must be unique across all subpckages of a package.
+	LocalDir string `yaml:"localDir,omitempty"`
+
+	// Upstream is a reference to where the subpackage should be fetched from.
+	// Whether a subpackage is local or remote is determined by whether Upstream is specified.
+	Upstream *Upstream `yaml:"upstream,omitempty"`
```

Comment:
It seems like we have two different definitions for subpackage and root package definition.  Upstream vs UpstreamLock.  Is the update strategy the difference maker here?  Why have two different ways and names to describe upstream info?

Reply:
`Upstream` is specified, `UpstreamLock` is resolved by kpt. There's an asymmetry that there's no `Upstream` for a top-level package, it's purely CLI-driven.

Reply:
Questions:
Not sure I understand what UpstreamLock is resolved by kpt mean?
Why is it called Lock?  Is it locking something?

Reply:
So the user's intent is given by `Upstream`. kpt consumes `Upstream` and manages `UpstreamLock` with the commit SHA-1 which fully "locks" the the Git commit that was fetched. The only time `UpstreamLock` changes is when:
a) User updates `Upstream` for a sub-package OR
b) User invokes `kpt cfg update` for the top-level package.